File: c/dining.c
#include <stdio.h>
#include <pthread.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include <unistd.h>
#include <semaphore.h>

#define NUM_PHILOSOPHERS 5
#define WAIT_TIME_MAX 3
#define WAIT_TIME_MIN 1

#define philosopher pthread_t
#define chopstick pthread_mutex_t

philosopher philosophers[NUM_PHILOSOPHERS];
chopstick chopsticks[NUM_PHILOSOPHERS];

// Think for a random amount of time between 1 and 3 seconds
void think(int id){
	// Think for a random amount of time
	printf("Philosopher %u is thinking...\n", id);
	sleep(rand() % WAIT_TIME_MAX + WAIT_TIME_MIN);
}


// Eat for a random amount of time between 1 and 3 seconds
void eat(int id){
	// Array to store our chopstick id's
	int c[2];
	int index;
	
	// For now lets assign both hands to one chopstick
	c[0] = id;
	c[1] = id;
	
	// Randomly choose which hand will grab the other chopstick
	c[rand() % 2] = (id + 1) % NUM_PHILOSOPHERS;

	// Attempt to grab chopsticks and do so until we have both of them
	for(index = 0; index < 2; index++){
		pthread_mutex_lock(&chopsticks[c[index]]);
		//printf("Philosopher %u has chopstick %u...\n", id, c[index]);
		//sleep(1);
	}
	
	// Print that we have both chopsticks
	printf("Philosopher %u has chopstick %u and %u.\n", id, c[0], c[1]);	
	
	// Eat
	sleep(rand() % WAIT_TIME_MAX + WAIT_TIME_MIN);

	// Release both chopsticks
	for(index = 0; index < 2; index++){
		pthread_mutex_unlock(&chopsticks[c[index]]);
	}
}


// The function all philosophers will run
void * run(void * arguments){
	// Grab arguments
	int argument = *((int *) arguments);
	
	// In an endless cycle think and then eat.
	while(1){
		think(argument);
		eat(argument);
	}
}


int main(int argc, char * argv[]){
	// Set up the seed for random numbers
	srand(time(NULL));
	
	// The array our arguments will be stored in
	// Just an array for philosopher id's
	int arguments[NUM_PHILOSOPHERS];
	int index;
	int result;

	// Initialize the Mutex
	for(index = 0; index < NUM_PHILOSOPHERS; index++){
		pthread_mutex_init(&chopsticks[index], NULL);
	}
	
	// Start each thread with the function and pass parameters to it
	for(index = 0; index < NUM_PHILOSOPHERS; index++){
		arguments[index] = index;
		result = pthread_create(&philosophers[index], NULL, run, &arguments[index]);		
	}

	// Join all threads
	for(index = 0; index < NUM_PHILOSOPHERS; index++){
		result = pthread_join(philosophers[index], NULL);
	}

	return 0;	
}

File: python/dining.py
# An implementation that solves the Dining
# Philosopher's problem using the Resource Hierarchy
# Solution proposed by Djikstra
import threading
import os
import time
from random import randint

# Number of philosophers as well as chopsticks
gNumPhilosophers = 5

# Lists of the philosophers(threads) and the chopsticks(locks)
gPhilosophers = []
gChopsticks = []

class Philosopher(threading.Thread):
	def __init__(self, id):
		threading.Thread.__init__(self)
		# Identifier used for printing the philosopher number
		self.id = id

	def run(self):
		# Assign left and right chopstick
		# Table visualization:
		leftChopstick = self.id
		rightChopstick = (self.id + 1) % gNumPhilosophers
		chopsticks = ChopstickPair(leftChopstick, rightChopstick)

		# Alternate between eating and 
		# thinking an arbitrary amount of time
		while True:
			# Think for some amount of time
			# between 0 and 15 seconds
			thinkingTime = randint(0, 15)
			time.sleep(thinkingTime)
			print "Philosopher {} finished thinking after {} second(s).".format(
				self.id, thinkingTime)

			# Attempt to eat
			# Keep track of how long the Philosopher waits
			beginWaitTime = time.time()
			chopsticks.pickUp()
			endWaitTime = time.time()
			
			totalWaitTime = endWaitTime - beginWaitTime

			# Eat for a short amount of time
			eatingTime = randint(0, 10)
			time.sleep(eatingTime)
			print "Philosopher {} ate for {} seconds. (Waited: {})".format(
				self.id, eatingTime, totalWaitTime)
			chopsticks.putDown()

			#repeat

# A class that defines a pair of chopsticks, requiring the index of the left
# chopstick to be less than the index of the right. This prevents deadlock.
class ChopstickPair:
	def __init__(self, leftChopstickIdx, rightChopstickIdx):
		# Order chopsticks by index to prevent deadlock
		if leftChopstickIdx > rightChopstickIdx:
			leftChopstickIdx, rightChopstickIdx = rightChopstickIdx, leftChopstickIdx
		self.firstChopstick = gChopsticks[leftChopstickIdx]
		self.secondChopstick = gChopsticks[rightChopstickIdx]

	def pickUp(self):
		# Acquire both chopstick locks by starting with lower idx
		self.firstChopstick.acquire()
		self.secondChopstick.acquire()

	def putDown(self):
		# Release the locks in any order
		self.firstChopstick.release()
		self.secondChopstick.release()


if __name__ == "__main__":
	# Initialize philosophers and chopsticks
	for i in range(gNumPhilosophers):
		gPhilosophers.append(Philosopher(i))
		gChopsticks.append(threading.Lock())

	# Start all philosophers thinking/eating cycle
	for philosopher in gPhilosophers:
		philosopher.start()

	# Allow CTRL + C to exit
	try:
		while True: time.sleep(0.1)
	except(KeyboardInterrupt, SystemExit):
		os._exit(0)
